### PR TITLE
Update the overview of essential classes

### PR DESCRIPTION
**Pull request by idorax**
fixes: #1559

### CODE REVIEW

**Comment by psss** [comment]
Looks good, could you please include classes inheriting from `SerializableContainer` such as `FmfId` or `RunData` as well?

**Comment by idorax** [comment]
> Looks good, could you please include classes inheriting from `SerializableContainer` such as `FmfId` or `RunData` as well?

`DataContainer` part is added. @psss, please help to review again, thanks!

**Comment by psss** [comment]
Yes, I agree that the list too long for a nice overview which would help to better understand the structure. I was thinking about separating the page into several sections which would be focused on one area:


Basic

    - tmt.Tree
    - tmt.base.Core:
      - tmt.Plan
      - tmt.Story
      - tmt.Test
    - tmt.Run
    - tmt.Clean
    - tmt.Status
    - tmt.Result

    - tmt.utils.SpecBasedContainer:
      - tmt.base.FmfId:
        - tmt.base.RequireFmfId
      - tmt.base.Link
      - tmt.base.Links

Steps

    - tmt.steps.Step:
      - tmt.steps.discover.Discover
      - tmt.steps.execute.Execute
      - tmt.steps.finish.Finish
      - tmt.steps.prepare.Prepare
      - tmt.steps.provision.Provision
      - tmt.steps.report.Report

      - tmt.steps.StepData:
        - tmt.steps.discover.DiscoverStepData:
          - tmt.steps.discover.fmf.DiscoverFmfStepData
          - tmt.steps.discover.shell.DiscoverShellData
        - tmt.steps.execute.ExecuteStepData:
          - tmt.steps.execute.internal.ExecuteInternalData
          - tmt.steps.execute.upgrade.ExecuteUpgradeData
        - tmt.steps.finish.FinishStepData:
          - tmt.steps.finish.shell.FinishShellData
        - tmt.steps.prepare.PrepareStepData:
          - tmt.steps.prepare.ansible.PrepareAnsibleData
          - tmt.steps.prepare.install.PrepareInstallData
          - tmt.steps.prepare.multihost.PrepareMultihostData
          - tmt.steps.prepare.shell.PrepareShellData
        - tmt.steps.provision.artemis.ProvisionArtemisData
        - tmt.steps.provision.connect.ProvisionConnectData
        - tmt.steps.provision.local.ProvisionLocalData
        - tmt.steps.provision.podman.ProvisionPodmanData
        - tmt.steps.provision.testcloud.ProvisionTestcloudData
        - tmt.steps.report.html.ReportHtmlData
        - tmt.steps.report.junit.ReportJUnitData
        - tmt.steps.report.polarion.ReportPolarionData

Phases

    - tmt.steps.Phase:
      - tmt.steps.Action:
        - tmt.steps.Login
        - tmt.steps.Reboot
      - tmt.steps.BasePlugin:
        - tmt.steps.GuestlessPlugin:
          - tmt.steps.discover.DiscoverPlugin:
            - tmt.steps.discover.fmf.DiscoverFmf
            - tmt.steps.discover.shell.DiscoverShell
          - tmt.steps.provision.ProvisionPlugin:
            - tmt.steps.provision.artemis.ProvisionArtemis
            - tmt.steps.provision.connect.ProvisionConnect
            - tmt.steps.provision.local.ProvisionLocal
            - tmt.steps.provision.podman.ProvisionPodman
            - tmt.steps.provision.testcloud.ProvisionTestcloud
          - tmt.steps.report.ReportPlugin:
            - tmt.steps.report.display.ReportDisplay
            - tmt.steps.report.html.ReportHtml
            - tmt.steps.report.junit.ReportJUnit
            - tmt.steps.report.polarion.ReportPolarion
        - tmt.steps.Plugin:
          - tmt.steps.execute.ExecutePlugin:
            - tmt.steps.execute.internal.ExecuteInternal:
              - tmt.steps.execute.upgrade.ExecuteUpgrade
          - tmt.steps.finish.FinishPlugin:
            - tmt.steps.finish.ansible.FinishAnsible
            - tmt.steps.finish.shell.FinishShell
          - tmt.steps.prepare.PreparePlugin:
            - tmt.steps.prepare.ansible.PrepareAnsible:
              - tmt.steps.finish.ansible.FinishAnsible
            - tmt.steps.prepare.install.PrepareInstall
            - tmt.steps.prepare.multihost.PrepareMultihost
            - tmt.steps.prepare.shell.PrepareShell

Guests

    - tmt.Guest:
      - tmt.GuestSsh:
        - tmt.steps.provision.artemis.GuestArtemis
        - tmt.steps.provision.testcloud.GuestTestcloud
      - tmt.steps.provision.local.GuestLocal
      - tmt.steps.provision.podman.GuestContainer

      - tmt.steps.provision.GuestData:
        - tmt.steps.provision.GuestSshData:
          - tmt.steps.provision.artemis.ArtemisGuestData:
            - tmt.steps.provision.artemis.ProvisionArtemisData
          - tmt.steps.provision.connect.ConnectGuestData:
            - tmt.steps.provision.connect.ProvisionConnectData
          - tmt.steps.provision.testcloud.TestcloudGuestData:
            - tmt.steps.provision.testcloud.ProvisionTestcloudData
        - tmt.steps.provision.local.ProvisionLocalData
        - tmt.steps.provision.podman.PodmanGuestData:
          - tmt.steps.provision.podman.ProvisionPodmanData

What do you think?

**Comment by idorax** [comment]
> What do you think?

@psss, I'm with you. Now the doc is updated, please help to review again, thanks!

**Comment by happz** [comment]
@idorax if I may, I'd mess with your patch a bit: as @psss pointed out, `provision` and `report` steps did not have their own step-specific `StepData` subclass for their plugins to use, and @psss suggested we should add them. I did so in https://github.com/teemtee/tmt/pull/1633, merging that PR before yours would make the `StepData` more symmetric (and sort of final as well, no other plans for changes).

**Comment by psss** [comment]
The new `provision` and `report` step data classes have been merged in #1633.
@idorax, could you please update the `StepDate` section accordingly? Thanks.

**Comment by idorax** [comment]
> The new `provision` and `report` step data classes have been merged in #1633. @idorax, could you please update the `StepDate` section accordingly? Thanks.

`StepData` section is updated. @psss, please help to review again, thanks!

**Review by psss — CHANGES_REQUESTED** [review]
Thanks, looks better now! Added two suggestions. Also, what about removing `Common` from all sections except from the first one? We can mention at the beginnig that most of the classes inherit from it. It probably does not make sense to repeat it in each and every section.

File: docs/classes.rst [hunk 1]
```diff
@@ -8,51 +8,117 @@
...
+
+Common
+└── Step
+    ├── Discover
+    ├── Execute
+    ├── Finish
+    ├── Prepare
+    ├── Provision
+    ├── Report
+    └── StepData
```
[the remark this answers is not in the document]
Comment (idorax) [reply docs/classes.rst]: It's corrected.

File: docs/classes.rst [hunk 2]
```diff
@@ -8,51 +8,117 @@
...
-    │       └── ReportHtml
-    ├── Guest
-    │   ├── GuestContainer
-    │   ├── GuestLocal
-    │   ├── GuestMinute
-    │   └── GuestTestcloud
-    ├── Run
-    ├── Status
-    └── Clean
+``Basic``
```
[the remark this answers is not in the document]
Comment (idorax) [reply docs/classes.rst]: Sure, added.

File: docs/classes.rst [hunk 3]
```diff
@@ -8,52 +8,118 @@
 Class Structure
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Here's the overview of core classes::
+Here's the overview of core classes which includes four parts::
+
+I - Basic classes
```
Comment (psss) [inline docs/classes.rst:13]: Please, add `::` to correctly introduce the preformatted block.

File: docs/classes.rst [hunk 4]
```diff
@@ -8,52 +8,118 @@
 Class Structure
 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-Here's the overview of core classes::
+Here's the overview of core classes which includes four parts::
```
Comment (psss) [inline docs/classes.rst:11]:
```suggestion
Here's the overview of core classes which includes four parts.
```

**Pull request merged**